amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseCleanupManager.java

```diff
@@ -0,0 +1,192 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package software.amazon.kinesis.leases;
+
+import lombok.EqualsAndHashCode;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import software.amazon.awssdk.services.kinesis.KinesisAsyncClient;
+import software.amazon.awssdk.services.kinesis.model.ChildShard;
+import software.amazon.awssdk.services.kinesis.model.GetRecordsRequest;
+import software.amazon.awssdk.services.kinesis.model.GetRecordsResponse;
+import software.amazon.awssdk.services.kinesis.model.GetShardIteratorRequest;
+import software.amazon.awssdk.services.kinesis.model.GetShardIteratorResponse;
+import software.amazon.awssdk.services.kinesis.model.ResourceNotFoundException;
+import software.amazon.awssdk.services.kinesis.model.ShardIteratorType;
+import software.amazon.kinesis.common.FutureUtils;
+import software.amazon.kinesis.common.KinesisRequestsBuilder;
+import software.amazon.kinesis.common.StreamIdentifier;
+import software.amazon.kinesis.leases.exceptions.DependencyException;
+import software.amazon.kinesis.leases.exceptions.InvalidStateException;
+import software.amazon.kinesis.leases.exceptions.ProvisionedThroughputException;
+import software.amazon.kinesis.metrics.MetricsFactory;
+import software.amazon.kinesis.retrieval.AWSExceptionManager;
+import software.amazon.kinesis.retrieval.kpl.ExtendedSequenceNumber;
+
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+/**
+ * Helper class to cleanup of any expired/closed shard leases. It will cleanup leases periodically as defined by
+ * {@link LeaseManagementConfig#leaseCleanupIntervalMillis()} upon worker shutdown, following a re-shard event or
+ * a shard expiring from the service.
+ */
+@Slf4j
+@RequiredArgsConstructor
+@EqualsAndHashCode
+public class LeaseCleanupManager {
+    @NonNull
+    private final StreamIdentifier streamIdentifier;
+    @NonNull
+    private final LeaseCoordinator leaseCoordinator;
+    @NonNull
+    private final KinesisAsyncClient kinesisClient;
+    @NonNull
+    private final MetricsFactory metricsFactory;
+    @NonNull
+    private final Duration maxFutureWait;
+    private final boolean cleanupLeasesUponShardCompletion;
+    private final boolean shouldCleanupLeases;
```

Comment:
1. Made this application level instead of stream/shard level, so we won't need to worry about garbage collecting the garbage collectors in next revision. Added state of pending deleted leases.
2. I don't see a reason to have multiple intervals for garbage/completed shards and feel this might be unnecessary, since we won't know for sure if a shard is garbage/completed until we actually try to use GetRecords. Current version of KCL doesn't have any differentiation between when Completed/Garbage leases are GC'd, so I don't really see a use case for this.